File: components/open3dcomp/src/specificworker.py
#
# Copyright (C) 2019 by YOUR NAME HERE
#
#    This file is part of RoboComp
#
#    RoboComp is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    RoboComp is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
#
from math import isnan

from genericworker import *

# If RoboComp was compiled with Python bindings you can use InnerModel in Python
# sys.path.append('/opt/robocomp/lib')
# import librobocomp_qmat
# import librobocomp_osgviewer
# import librobocomp_innermodel
import open3d as o3
import cv2
import numpy as np
import traceback


class SpecificWorker(GenericWorker):
    def __init__(self, proxy_map):
        super(SpecificWorker, self).__init__(proxy_map)
        self.timer.timeout.connect(self.compute)
        self.Period = 2000
        self.timer.start(self.Period)

    def __del__(self):
        print 'SpecificWorker destructor'

    def setParams(self, params):
        # try:
        #	self.innermodel = InnerModel(params["InnerModelPath"])
        # except:
        #	traceback.print_exc()
        #	print "Error reading config params"
        return True

    @QtCore.Slot()
    def compute(self):
        try:
            # color, depth, _, _ = self.rgbd_proxy.getData()
            nube, _, _ = self.rgbd_proxy.getXYZ()
            colors, _, _ = self.rgbd_proxy.getRGB()
            print len(nube)
            self.pintar(nube, colors)
        # image = np.frombuffer(color, dtype=np.uint8)
        # print image.size
        # image = np.reshape(image, (480, 640, 3))
        # image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        # cv2.imshow("color", image)
        # print (min(depth), max(depth))
        # depth = np.interp(depth, [min(depth), max(depth)], [0, 255])
        # print(min(depth), max(depth))
        # depth = np.reshape(depth, (480, 640, 1))
        # cv2.imshow("depth", depth.astype(np.uint8))
        # cv2.waitKey(5)
        # print len(color)
        except Ice.Exception, e:
            traceback.print_exc()
            print e

        return True

    def pintar(self, nube, colors):
        #
        # points = [[0, 0, 0], [1, 0, 0], [0, 1, 0], [1, 1, 0],
        #           [0, 0, 1], [1, 0, 1], [0, 1, 1], [1, 1, 1]]
        # lines = [[0, 1], [0, 2], [1, 3], [2, 3],
        #          [4, 5], [4, 6], [5, 7], [6, 7],
        #          [0, 4], [1, 5], [2, 6], [3, 7]]
        # colors = [[1, 0, 0] for i in range(len(lines))]
        # line_set = o3.LineSet()
        # line_set.points = o3.Vector3dVector(points)
        # line_set.lines = o3.Vector2iVector(lines)
        # line_set.colors = o3.Vector3dVector(colors)
        # o3.draw_geometries([line_set])

        xyz, rgb = [], []
        print nube[10000:10100]
        # print colors[10000:10100]
        print type(nube)
        whites = 0
        nan_count = 0
        for p, c in zip(nube, colors):
            if not (isnan(p.x) or isnan(p.y) or isnan(p.z)):
                xyz.append([p.x, p.y, p.z])
                if not (isnan(c.red) or isnan(c.green) or isnan(c.blue)):
                    rgb.append([c.red, c.green, c.blue])
                else:
                    rgb.append([0, 0, 255])
                if c.red==255 and c.green==255 and c.blue==255:
                    whites+=1
            else:
                nan_count+=1
        print "Ostia que blanco " + str(whites)
        print "Ostia que NAN " + str(nan_count)

        # colors = [[1, 4, 0] for i in range(len(xyz))]
        # Pass xyz to Open3D.PointCloud and visualize
        pcd = o3.PointCloud()

        # pcd = o3.LineSet()
        pcd.points = o3.Vector3dVector(xyz)
        pcd.colors = o3.Vector3dVector(rgb)
        # downpcd = o3.voxel_down_sample(pcd, voxel_size=0.1)
        o3.draw_geometries([pcd])

        # print(np.asarray(pcd.points))
        # o3.draw_geometries([pcd])

    # print("Downsample the point cloud with a voxel of 0.05")

File: experimental/CGR/src/specificworker.cpp
//
//    Copyright (C) 2015 by YOUR NAME HERE
//
//    This file is part of RoboComp
//
//    RoboComp is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    RoboComp is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
//========================================================================
//
// based in Joydeep Biswas, (C) 2010 paper
//
//========================================================================
//


#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>

// #include <ros/ros.h>
// #include <tf/tf.h>
// #include <tf/transform_listener.h>
// #include <tf/transform_broadcaster.h>
// #include <sensor_msgs/LaserScan.h>
// #include <sensor_msgs/Image.h>
// #include <nav_msgs/Odometry.h>
// #include <geometry_msgs/PoseArray.h>
// #include <geometry_msgs/PoseWithCovarianceStamped.h>
// #include <ros/package.h>

//#include "popt_pp.h"
#include "proghelp.h"
//#include "cgr_localization/DisplayMsg.h"
//#include "cgr_localization/LocalizationInterfaceSrv.h"
//#include "cgr_localization/LocalizationMsg.h"

#include "vectorparticlefilter.h"
#include "vector_map.h"

#include "terminal_utils.h"
#include "timer.h"



#include "configreader.h"
//#include "plane_filtering.h"

#include "specificworker.h"

using namespace std;
/**
* \brief Default constructor
*/
SpecificWorker::SpecificWorker(MapPrx& mprx) : GenericWorker(mprx)
{
/*
    LoadParameters();
 
    printf("NumParticles     : %d\n",numParticles);
    printf("Alpha1           : %f\n",motionParams.Alpha1);
    printf("Alpha2           : %f\n",motionParams.Alpha2);
    printf("Alpha3           : %f\n",motionParams.Alpha3);
    printf("UsePointCloud    : %d\n",usePointCloud?1:0);
    printf("UseLIDAR         : %d\n",noLidar?0:1);
    printf("Visualizations   : %d\n",debugLevel>=0?1:0);
    printf("\n");
  
  double seed = floor(fmod(GetTimeSec()*1000000.0,1000000.0));
  if(debugLevel>-1) printf("Seeding with %d\n",(unsigned int)seed);
  srand(seed);
*/
 
  //Initialize particle filter, sensor model, motion model, refine model
  string mapsFolder("maps");
  localization = new VectorLocalization2D(mapsFolder.c_str());
	
}

/**
* \brief Default destructor
*/
SpecificWorker::~SpecificWorker()
{
	
}

void SpecificWorker::LoadParameters()
{

  WatchFiles watch_files;
  ConfigReader config("etc/params.conf");
  /*
  config.init(watch_files);
    
  {
    ConfigReader::SubTree c(config,"initialConditions");
    
    bool error = false;
    curMapName = string(c.getStr("mapName"));
    error = error || curMapName.length()==0;
    error = error || !c.getVec2f("loc",initialLoc);
    error = error || !c.getReal("angle", initialAngle);
    error = error || !c.getReal("locUncertainty", locUncertainty);
    error = error || !c.getReal("angleUncertainty", angleUncertainty);
    
    if(error){
      printf("Error Loading Initial Conditions!\n");
      exit(2);
    }
  }
  
  {
    ConfigReader::SubTree c(config,"motionParams");
    
    bool error = false;
    error = error || !c.getReal("Alpha1", motionParams.Alpha1);
    error = error || !c.getReal("Alpha2", motionParams.Alpha2);
    error = error || !c.getReal("Alpha3", motionParams.Alpha3);
    error = error || !c.getReal("kernelSize", motionParams.kernelSize);
    
    
    if(error){
      printf("Error Loading Predict Parameters!\n");
      exit(2);
    }
  }
  
  {
    ConfigReader::SubTree c(config,"lidarParams");
    
    bool error = false;
    // Laser sensor properties  //BASURA
    error = error || !c.getReal("angleResolution", lidarParams.angleResolution);
    error = error || !c.getInt("numRays", lidarParams.numRays);
    error = error || !c.getReal("maxRange", lidarParams.maxRange);
    error = error || !c.getReal("minRange", lidarParams.minRange);
    
    // Pose of laser sensor on robot	//BASURA
    vector2f laserToBaseTrans;
    float xRot, yRot, zRot;
    error = error || !c.getVec2f<vector2f>("laserToBaseTrans", laserToBaseTrans);
    error = error || !c.getReal("xRot", xRot);
    error = error || !c.getReal("yRot", yRot);
    error = error || !c.getReal("zRot", zRot);
    Matrix3f laserToBaseRot;
    laserToBaseRot = AngleAxisf(xRot, Vector3f::UnitX()) * AngleAxisf(yRot, Vector3f::UnitY()) * AngleAxisf(zRot, Vector3f::UnitZ());
    lidarParams.laserToBaseTrans = Vector2f(V2COMP(laserToBaseTrans));
    lidarParams.laserToBaseRot = laserToBaseRot.block(0,0,2,2);
    
    // Parameters related to observation update
    error = error || !c.getReal("logObstacleProb", lidarParams.logObstacleProb);
    error = error || !c.getReal("logOutOfRangeProb", lidarParams.logOutOfRangeProb);
    error = error || !c.getReal("logShortHitProb", lidarParams.logShortHitProb);
    error = error || !c.getReal("correlationFactor", lidarParams.correlationFactor);
    error = error || !c.getReal("lidarStdDev", lidarParams.lidarStdDev);
    error = error || !c.getReal("attractorRange", lidarParams.attractorRange);
    error = error || !c.getReal("kernelSize", lidarParams.kernelSize);
    
    // Parameters related to observation refine
    error = error || !c.getInt("minPoints", lidarParams.minPoints);
    error = error || !c.getInt("numSteps", lidarParams.numSteps);
    error = error || !c.getReal("etaAngle", lidarParams.etaAngle);
    error = error || !c.getReal("etaLoc", lidarParams.etaLoc);
    error = error || !c.getReal("maxAngleGradient", lidarParams.maxAngleGradient);
    error = error || !c.getReal("maxLocGradient", lidarParams.maxLocGradient);
    error = error || !c.getReal("minCosAngleError", lidarParams.minCosAngleError);
    error = error || !c.getReal("correspondenceMargin", lidarParams.correspondenceMargin);
    error = error || !c.getReal("minRefineFraction", lidarParams.minRefineFraction);
    
    lidarParams.initialize();
    
    if(error){
      printf("Error Loading Lidar Parameters!\n");
      exit(2);
    }
  }
*/
}

bool SpecificWorker::setParams(RoboCompCommonBehavior::ParameterList params)
{
//       THE FOLLOWING IS JUST AN EXAMPLE
//
// 	try
// 	{
// 		RoboCompCommonBehavior::Parameter par = params.at("InnerModelPath");
// 		innermodel_path=par.value;
// 		innermodel = new InnerModel(innermodel_path);
// 	}
// 	catch(std::exception e) { qFatal("Error reading config params"); }
	
	
	timer.start(Period);

	return true;
}


void SpecificWorker::compute()
{
//static RoboCompLaser::TLaserData laserData;

 		//rgbd_proxy->getXYZ(points, hState, bState);
	//	laserData = laser_proxy->getLaserData();
	//		int j=0;
	//		for(auto i : laserData)
	//	{
			//pasar a xyz añadiendo w QVec p1 = QVec::vec3(x,y,z); 
			//points[j].x=0;
			//points[j].y=0;
			//points[j].z=0;
			//points[j].w=0;
			//j++;
	//	}
// 	try
// 	{
// 		camera_proxy->getYImage(0,img, cState, bState);
// 		memcpy(image_gray.data, &img[0], m_width*m_height*sizeof(uchar));
// 		searchTags(image_gray);
// 	}
// 	catch(const Ice::Exception &e)
// 	{
// 		std::cout << "Error reading from Camera" << e << std::endl;
// 	}
}





